Location delete

calellowitz opened the pull request:
Adding the ability to delete locations! also deletes all child locations, unassigns users, and closes supply point cases. Deletion fails when custom models have a foreign key that points to location being deleted, but archives stock states that do the same.
@esoergel cc: @sravfeyn @kaapstorm 
Can probably be reviewed all at once as most changes are self-contained to a single file.

calellowitz committed:
adding location delete view and model functions

calellowitz committed:
adding location delete fronted

calellowitz committed:
preserving stock state objects when linked location is deleted

calellowitz committed:
making location delete fail if location linked to custom model

calellowitz committed:
changing location delete to not override normal method

calellowitz committed:
getting custom tests with locations passing

calellowitz committed:
getting existing commtrack tests passing again

calellowitz committed:
changing StockState manager to exclude stock states with deleted location

calellowitz committed:
adding tests for location delete

calellowitz committed:
changing location delete to use atomic transaction

calellowitz committed:
changing delete to use bulk operation and minor refactoring

calellowitz committed:
improving ui for failed deletes

calellowitz committed:
adding tests to ensure proper handling of foreign key relationships

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -360,6 +360,20 @@ def archive_location(request, domain, loc_id):
         )
     })
 
+@can_edit_location
```
E302 expected 2 blank lines, found 1

dimagimon commented on custom/ewsghana/tests/test_delete_domain.py:
```diff
@@ -58,3 +58,4 @@ def test_pre_delete_signal_receiver(self):
 
     def tearDown(self):
         self.domain2.delete()
+
```
W391 blank line at end of file

dimagimon commented on custom/ewsghana/tests/test_locations_sync.py:
```diff
@@ -204,3 +205,15 @@ def test_facility_in_charge(self):
+        with self.assertRaises(ProtectedError):
+            ewsghana_location.full_delete()
+        ewsghana_sql_location = SQLLocation.objects.filter(location_id=ewsghana_location._id)
+        self.assertEqual(len(ewsghana_sql_location), 1)
```
W292 no newline at end of file

calellowitz commented on corehq/apps/locations/templates/locations/manage/locations.html:
```diff
@@ -264,4 +310,22 @@
+            {% blocktrans %}
+            <strong>Warning!</strong> Deleting a location is permanent. There is no way to undo this action.
+            This will also delete all child locations. Are you sure you want to delete this location?
+            {% endblocktrans %}
```
@dimagi/product if you have feedback on the warning language, i would be happy to change it.

sravfeyn replied:
I think  it might be helpful to append that if those locations owns cases, they will be closed.

esoergel replied:
It would also be nice to show how many child locations will be deleted.  Something like:
```django
{% if not num_children %}
You are about to delete a location.
{% else %}
You are about to delete this location and its {{ num_children }} child locations.
{% endif %}
Are you sure....
```

sravfeyn commented on corehq/apps/commtrack/tests/util.py:
```diff
@@ -164,6 +164,7 @@ def setUp(self):
 
         self.loc = make_loc('loc1')
         self.sp = make_supply_point(self.domain.name, self.loc)
+        self.loc.save()
```
Why does `loc` need to be saved again? `make_loc` saves it, right?

calellowitz replied:
the signal that links the stock state created by make_supply_point to the location in the sql database is only fired when the location is saved. when i changed the manager for the stockstate model to exclude stock states with no locations, i needed to get ensure they would be properly linked and saving the location again was the only way to do that.

sravfeyn replied:
Cool, makes sense.

Does this mean that the location should be saved in each place `make_supply_point ` is called? Or may be signal hooks should be modified to take care of this, perhaps `sync_location_supply_point` should be triggered after `make_supply_point` as well?  I am not super familiar with this, @esorgel might have ideas.

esoergel replied:
I don't have much more context on this than you two, but it seems reasonable to fire the `location_edited` signal in `make_supply_point`.

sravfeyn commented on corehq/apps/locations/models.py:
```diff
@@ -588,6 +581,37 @@ def unarchive(self):
+
+    def full_delete(self):
+        """
+        Delete a location and its dependants.
```
What happens if there are more than two levels of dependents? Shouldn't that trigger a chain until there are no dependants at any level?

calellowitz replied:
self.descendants should get all levels of children, and each of the locations returned from that have their couch doc deleted. the sql deletes automatically cascade down when a child has a foreign key relationship with a deleted parent, which is why there is only one sql delete call.

sravfeyn replied:
Ha, makes sense!

esoergel replied:
Just realized there's a typo - `dependants != descendants`

calellowitz replied:
good catch! although to be fair in this case, those locations are descendants and (in some ways) dependants

sravfeyn commented on corehq/apps/locations/templates/locations/manage/locations.html:
```diff
@@ -81,6 +92,37 @@
+    function delete_fn() {
+        $(button).disableButton();
+        $.ajax({
+            type: 'POST',
```
Should this be `DELETE`?

sravfeyn commented on corehq/apps/locations/views.py:
```diff
@@ -360,6 +360,20 @@ def archive_location(request, domain, loc_id):
     })
 
+@can_edit_location
+def delete_location(request, domain, loc_id):
```
Does this need `method.request` check? Otherwise seems like GET to /<loc_id>/delete might be able to delete too?

esoergel replied:
You should be able to do this: https://github.com/dimagi/commcare-hq/blob/91f8a3e6fbefc223a411773020cabd0556c4a140/corehq/apps/reports/views.py#L723  (That's how the django decorators `@require_POST` and `@require_GET` are implemented)

sravfeyn commented:
Build is failing below.
````python
======================================================================
FAIL: test_balance_submission (corehq.form_processor.tests.test_ledgers.LedgerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/util/test_utils.py", line 174, in inner
    return helper(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/util/test_utils.py", line 147, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/util/test_utils.py", line 162, in call_with_settings
    fn(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/util/test_utils.py", line 143, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/tests/test_ledgers.py", line 49, in test_balance_submission
    self._assert_ledger_state(100)
  File "/home/travis/build/dimagi/commcare-hq/corehq/form_processor/tests/test_ledgers.py", line 78, in _assert_ledger_state
    self.assertEqual(1, len(ledgers))
AssertionError: 1 != 0
----------------------------------------------------------------------

sravfeyn commented:
Looks good, made few comments. Does this need QA?

calellowitz commented:
@sravfeyn qa ticket is already created. i will look into the build fail and the comments i haven't yet responded to tomorrow.

kaapstorm commented on corehq/apps/locations/models.py:
```diff
@@ -588,6 +581,37 @@ def unarchive(self):
+        Delete a location and its dependants.
+        This also unassigns users assigned to the location.
+        """
+        to_delete = [loc for loc in [self] + self.descendants]
```
Small, but self.descendants is already a list, so this could be `to_delete = [self] + self.descendants` (just as you do in line 581) without needing to iterate them here, and then again in line 611.

esoergel commented on corehq/apps/commtrack/tests/test_locations.py:
```diff
@@ -87,6 +87,43 @@ def test_archive_flips_sp_cases(self):
+        )
+        test_loc.save()
+
+        original_count = len(list(Location.by_domain(self.domain.name)))
```
Is this part necessary?  The rest of the test assumes there's only the user's location and test_loc.

esoergel commented on corehq/apps/commtrack/models.py:
```diff
@@ -410,13 +410,14 @@ def get_display_config(cls):
             .exclude(sql_product__is_archived=True) \
-            .exclude(sql_location__is_archived=True)
+            .exclude(sql_location__is_archived=True) \
+            .exclude(sql_location=None)
```
I would check to see that no existing StockStates have `sql_location == None`, as this might change behavior unexpectedly.

You might also be able to just delete these StockStates.  @sheelio What do you think?  In my opinion, if everything is deleted reversibly, then how is that different than archiving?

calellowitz replied:
I talked to Mike O about this and no one has been able to think of a situation where you can have a stock state without a location, so there should not be any harm in this. Since those states are primarily tied to supply point cases, and we keep the cases around this seemed the better option, but I would be happy to delete those too, if that is what people think is better.

esoergel replied:
I just ran this on prod:
```python
>>> StockState.objects.filter(sql_location=None).count()
133736
```
I don't know how those were created, but that is something we'll need to account for.  I agree with your sentiment that we should do with these whatever we do with cases, but I do think in this case it might be better to just delete the stock states and mention that in the warning when you try to delete a location.

esoergel replied:
Oh, I just saw your comment below - looks like we're already on the same page here, feel free to disregard. :smile:

esoergel commented on corehq/apps/locations/models.py:
```diff
@@ -540,14 +540,7 @@ def _archive_single_location(self):
-            close_case(sp.case_id, self.domain, COMMTRACK_USERNAME)
-
-        _unassign_users_from_location(self.domain, self._id)
+        self._close_case_and_remove_users()
```
:+1:

sheelio commented:
Seems okay by me to delete stock states (if you're truly deleting).

calellowitz committed:
fixing pr feedback

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -360,6 +360,21 @@ def archive_location(request, domain, loc_id):
         )
     })
 
+@require_http_methods(['DELETE'])
```
E302 expected 2 blank lines, found 1

calellowitz committed:
deleting linked stock states and including those with no location

calellowitz committed:
fixing documentation for stockstate manager

calellowitz commented:
@esoergel @sravfeyn @sheelio 
I changed behavior to deleting stock states and stopped filtering stock states with deleted locations because in addition to breaking a lot of tests, I looked at the prod database and it looked like about half of the stock states in there had no attached location, including many that were last modified today. A simple delete while leaving manager behavior unchanged seemed like the right way forward. It should also get all tests passing again.

sravfeyn commented:
Looks good to me but the test `corehq.apps.commtrack.tests.test_stock_state.StockStateBehaviorTest` seems to be failing

calellowitz committed:
addressing more pr feedback

snopoke committed:
Merge branch 'master' into location-delete

snopoke committed:
merge migarations from master

calellowitz committed:
Merge branch 'master' into location-delete

calellowitz committed:
fixing migration conflict

calellowitz commented:
QA Passed on this. This is good to go unless @dimagi/product wants to hold off.

czue commented:
sorry for the last minute jump in - have had this PR open for a long time, but didn't think there was time pressure to review till now! this feature seems a little scary given that it's an irrecoverable deletion, and just want to make sure we've completely thought through / tested.

Can you quickly answer:

- Did you consider adding a type-something-in-a-box type of warning for this similar to deleting users?
- Did you consider changing the source doc type to `Location-Deleted` rather than hard deleting them from the DB?
- What happens to cases that are owned by a deleted location?
- Have we thought through the implications of what happens to phones that still have location case-sharing groups on them? What happens when they submit data associated with the location?
- Hopefully covered in QA, but how does this affect historical data owned by the location? Assume still usable somehow in reports?

(I may have more questions, but will start there for now)

czue commented:
Hope you don't mind, but I put a "don't merge" label on this too just until you get a chance to respond.

amsagoff commented:
Hey - I'm not sure how to review this from a product perspective (or if
someone like sheel already has). Is there a spec for this?

On Wed, Jan 20, 2016 at 11:14 AM, Cory Zue <notifications@github.com> wrote:

> Hope you don't mind, but I put a "don't merge" label on this too just
> until you get a chance to respond.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/9618#issuecomment-173253770>.
>

esoergel commented on corehq/apps/locations/templates/locations/manage/locations.html:
```diff
@@ -175,6 +217,10 @@
+                        <button class="btn btn-danger"
+                            data-bind="click: function(data, event) { delete_loc(event.currentTarget, name(), uuid()) }, visible: show_archive_action_button">
+                            {% trans "Delete" %}
+                        </button>
```
Might make sense to show this also on the `show_inactive` version, that is - move it outside this `{% else %}` block.

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into location-delete

esoergel committed:
Delete SQLLocations after cases

This just makes the merge with my feature work properly (as getting
cases requires sql_locations).
@calellowitz FYI

sheelio commented:
@czue @esoergel @amsagoff 

Ethan and I talked about this a fair bit before Cal started
- It definitely makes sense to pop up a warning about the irreversible-ness of the delete and force you to confirm, a-la mobile worker delete
- I think Ethan was nervous about the LOE of making something more reversible.  Even if we do a delete from SQL, I think it still makes sense to have the doc hang around with a changed doc-type so we can restore the location if really needed
- delete behaviour - its going to be similar to what happens if you delete a group on HQ (and cases are owned by that group, or more cases are submitted against that owner ID).  Cases will be in this weird "unknown owner" state, and I think future cases submitted will also be in that state.  

I don't think its important to make the location or its data usable if you delete it - we're treating it as equivalent to deleting a mobile worker. 

If we do care about keeping the location usable, I think the right option is to change the definition of "archive" to be "unassigning that location from a parent".  This way the data is still attached to a known location, but you can move and delete locations without worrying about associated archived locations.  

(can go to voice to further discuss as well).

esoergel commented:
regarding `Location-Deleted`:
I really don't think we should build any new functionality on the particulars of couch locations, as I'm trying to get rid of them.

Also, we already have an "Archive" button for locations, so I don't think it makes sense to also have a reversible delete. The reason "Archive" sucks is because it doesn't actually delete anything - in order to be able to restore the location, you need to keep its location code unique, and you need to keep around a relationship from it to its parent, its location type, and its stock_states.  This makes it hard to change any of those things in the future.

I think what people want to be able to do is say "I messed this up, let me delete it and try again", and that doesn't fit well with archival.

If we could come up with a more usable archival workflow, then I don't think we'd need to be able to delete them, but I think that's a much harder problem.

calellowitz commented:
I think Sheel and Ethan have a lot more context here than I do and answered most your questions, so I will defer to them on most of the substance. In general, my understanding was that this was only to be used for drastic restructuring and so data associated with previous structures are less relevant, legacies of the past structure are what we are trying to avoid. I think better warning to that effect makes a lot of sense.
Regarding the type to confirm box, would you want to require them to type all descendants as well? It seems misleading to allow them to delete while only confirming one, but confirming each could be difficult if the location has a lot.

czue commented:
Okay makes sense. I had forgotten that we already have to deal with a lot of these things for groups. It kind of seems to me like we just got archival wrong and I wonder if we should revisit that at all, but is a conversation for a later time.

> Regarding the type to confirm box, would you want to require them to type all descendants as well? It seems misleading to allow them to delete while only confirming one, but confirming each could be difficult if the location has a lot

@calellowitz agree it shouldn't be one-per, although I also like making the user indicate that they have parsed what we are showing them and understand that they are losing all the children. What if we said something like:

"This will delete the location and all 38 of its descendants. To confirm that you understand this please enter the total number of locations that will be deleted"

And then they have to type in "39". Too cute?

czue commented:
thanks for quick responses everyone. I think just need to add some flavor of the delete confirmation and then this is good to go! :rocket:

esoergel commented:
+1 on showing the descendants count - that's just one more warning that they might be doing something they don't intend.

amsagoff commented:
Let me know if helpful for me to be involved and I'll take a look at the
feature - sounds like y'all are aligned so happy to sit out too

On Thu, Jan 21, 2016 at 10:08 AM, Ethan Soergel <notifications@github.com>
wrote:

> +1 on showing the descendants count - that's just one more warning that
> they might be doing something they don't intend.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/9618#issuecomment-173599993>.
>

calellowitz committed:
adding type to confirm for location delete

dimagimon commented on corehq/apps/locations/views.py:
```diff
@@ -360,6 +360,30 @@ def archive_location(request, domain, loc_id):
+        )
+    })
+
+def location_descendants_count(request, domain, loc_id):
```
E302 expected 2 blank lines, found 1

calellowitz committed:
[ci skip] lint errors

calellowitz commented:
New warning modal
<img width="636" alt="screen shot 2016-01-22 at 12 25 35 pm" src="https://cloud.githubusercontent.com/assets/6844721/12517933/644daab6-c103-11e5-982d-75bb20e6aa37.png">
<img width="698" alt="screen shot 2016-01-22 at 12 25 44 pm" src="https://cloud.githubusercontent.com/assets/6844721/12517938/686c8e00-c103-11e5-880a-d549cd31097c.png">

esoergel commented:
Looks good to me.  Is the `Open for review: hold off on merge` tag still on there for a reason?

calellowitz commented:
@czue added that. I was going to wait for him to take it off but he seemed only concerned about the warning message

esoergel commented:
Okay, I think this is g2g then.  Merging.

czue commented:
@calellowitz thanks - I was actually proposing we not tell them what number to type in, so instead of "please type **7**" it would say "please type the total number of locations" and they would have to parse the sentence well enough to realize that it should be the number of descendants plus the location itself.

Also I think the text could be tweaked to clarify where the "7" is coming from. My proposal was something like this:

```
Warning!  Deleting a location is permanent. There is no way to undo this action.
This will also delete all child locations. If you are sure you would like to delete this location 
and all 7 of its children, please type the total number of locations to be deleted below.
```

I'm open to push back from you or @dimagi/product on the addition part if that seems like overkill. Another option would be to have them type the location name.

czue commented:
either way, exciting to see this merged!
